src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,386 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+                    //    ]
+                    //}
+
+                    // If we have a request payload above and we are handling the changed values in Orders,
+                    // The apiHandlerOfT is OrdersAPIHandler with parent as Customers(1).
+                    // The object with "@odata.id" will have OrdersAPIHandler with parent as Customers(3).
+                    // The object with id 2000 will have OrdersAPIHandler similar to apiHandlerOfT.
+                    // The codebelow ensures we use the correct handler.
+                    // NOTE: Handler when we have odata.id will always be different than the parent handler.
+                    // However this is true if handlers are applied correctly.
+
+                    ODataAPIHandler<TStructuralType> odataPathApiHandler = apiHandlerFactory.GetHandler(changedObj.ODataPath) as ODataAPIHandler<TStructuralType>;
+
+                    if (odataPathApiHandler != null && changedObj.ODataPath.Any())
+                    {
+                        // Using a 3rd party library to compare objects
+                        // https://github.com/GregFinzer/Compare-Net-Objects/wiki/Getting-Started
+                        // TODO:
+                        // a) Confirm to perf regressions are caused byt his library.
+                        // b) Investigate if there is a better library.
+
+                        KellermanSoftware.CompareNetObjects.CompareLogic compare = new KellermanSoftware.CompareNetObjects.CompareLogic();
+                        compare.Config.CaseSensitive = false;
+                        compare.Config.ComparePrivateProperties = true;
+                        compare.Config.ComparePrivateFields = true;
+
+                        KellermanSoftware.CompareNetObjects.ComparisonResult result = compare.Compare(handler, odataPathApiHandler);
+
+                        if (!result.AreEqual)
+                        {
+                            handler = odataPathApiHandler;
+                            hasODataId = true;
+                        }
+                    }
+                }
+
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                Dictionary<string, object> keyValues = new Dictionary<string, object>();
+
+                foreach (string key in _keys)
+                {
+                    object value;
+
+                    if (changedObj.TryGetPropertyValue(key, out value))                   
+                    {
+                        keyValues.Add(key, value);
+                    }
+                }
+
+                try
+                {
+                    if (hasODataId)
+                    {
+                        keyValues = changedObj.ODataPath.GetKeys();
```

Comment:
I don't think this method returns the correct Id in the case of an ODataId.. Can we get the Id from ODataIdContainer..

`Books(2)/Presses`

and we have `@odata.id:Presses(9)`..The GetKeys method will return 2 and not 9

Reply:
`Books(2)/Presses` is the `DeltaSet<T>`.

`changedObj` is the `Delta<T>` so the odata path string will be `Presses(9)`.

Reply:
Resolved offline.